
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Data model &#8212; immuneSIM 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="data-model">
<h1>Data model<a class="headerlink" href="#data-model" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<p>The ImmuneML’s data model consists of:</p>
<ul class="simple">
<li>Sequence class</li>
<li>Repertoire class</li>
<li>Dataset class</li>
</ul>
<div class="section" id="sequence">
<h2>Sequence<a class="headerlink" href="#sequence" title="Permalink to this headline">¶</a></h2>
<p>The Sequence class contains all information about a single immune receptor sequence. For example, a sequence can refer to
a CDR3 sequence. The information stored about a sequence are:</p>
<ul class="simple">
<li>amino acid sequence,</li>
<li>nucleotide sequence,</li>
<li>id,</li>
<li>annotation,</li>
<li>metadata,</li>
</ul>
<p><strong>Id</strong> is any unique string that will unambiguously identify the sequence.</p>
<p><strong>Sequence metadata</strong> contains information about V gene, J gene, chain, sequence count and frame type. They can be used for
further analysis. For instance, when analyzing sequences, those that have “Out” or “Stop” frame types would be discarded.</p>
<p><strong>Sequence annotation</strong> is used for simulation purposes. In cases when sequences are modified to include an artificial
disease signal, this object will store the information on how exactly the sequence was modified.</p>
</div>
<div class="section" id="repertoire">
<h2>Repertoire<a class="headerlink" href="#repertoire" title="Permalink to this headline">¶</a></h2>
<p>The Repertoire class contains information about a single immune repertoire. A repertoire object consists of:</p>
<ul class="simple">
<li>a list of sequences and</li>
<li>a metadata object.</li>
</ul>
<p>A list of <strong>sequences</strong> includes all sequences coming from the same person. Each sequence is an instance of a <span class="xref std std-ref">Sequence</span> object.</p>
<p><strong>Repertoire metadata</strong> contains information about a repertoire. That information is modeled by information about a sample
and, in case of simulation, list of modifications to sequences in the repertoire.</p>
<p><strong>Sample</strong> is defined by a unique identifier, an optional name (in case the unique identifier is not descriptive enough)
and optional other parameters. Examples of such parameters could be:</p>
<ul class="simple">
<li>date of the experiment,</li>
<li>age of the patient,</li>
<li>known diseases of the patient etc.</li>
</ul>
<p>In case only sequences should be analyzed, regardless of the repertoires, the <em>Sequence</em>, <em>Repertoire</em> and <em>Dataset</em> classess
should still be used, but it is necessary then to make each repertoire to consist of only one sequence. Everything else in the
analysis, except where noted, can be used in the same manner as when the analysis has repertoires consisting of a bulk of
sequences.</p>
</div>
<div class="section" id="dataset">
<h2>Dataset<a class="headerlink" href="#dataset" title="Permalink to this headline">¶</a></h2>
<p>Dataset class models a list of repertoires. It contains the following information:</p>
<ul class="simple">
<li>a unique identifier,</li>
<li>a list of repertoires,</li>
<li>a list of filenames,</li>
<li>encoded repertoires and</li>
<li>dataset parameters.</li>
</ul>
<p>If not set manually by the user, the unique identifier is automatically generated.</p>
<p>A list of repertoires is a list of objects of <span class="xref std std-ref">Repertoire</span> class. In case the repertoires occupy too much memory,
and cannot be loaded all at once, the dataset contains a list of paths to each repertoire file for the dataset. Each
repertoire then is loaded from the file as needed, thus avoiding memory issues.</p>
<p>Encoded repertoires are used for machine learning setting. Since machine learning algorithms cannot work with the data
in their original format, they are encoded so that they can be further analyzed. Examples of the encoding include k-mer
decomposition and encoded a repertoire by k-mer frequencies, vector embeddings on repertoire level and others.</p>
<p>Dataset parameters are an instance of <strong>DatasetParams</strong> class and include the following:</p>
<ul class="simple">
<li>number of repertoires in the dataset,</li>
<li>path to dataset file,</li>
<li>name of the encoding type,</li>
<li>a list of parameters available in samples for each repertoire (e.g. date, age, disease).</li>
</ul>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">immuneSIM</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, CRW.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.5</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/parameters.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>